### ISSUE TITLE
Weird encoding issue when using a certain type of network

### ISSUE THREAD

**Issue by catb0t** [factor/factor#1589]
So, [I posted this on Stack Overflow to see what I could find](http://stackoverflow.com/q/36230157), but because I'm so sure this is unintentional, I'll add it here as a maybe-bug-or-just odd thing. The SO question copy-pasted:

---
I've a smartphone. On this smartphone, I've a mobile hotspot, essentially a portable WiFi network that pipes my phone's internet access to my laptop.

On my laptop, I've Python 3 and the [requests library](http://docs.python-requests.org). Here's using Python and requests to get `google.com`, with my phone's hotspot. (result is exactly the same using "real wifi".)

    >>> x = requests.get("http://google.com")
    >>> x.apparent_encoding; x[:100]
    'ISO-8859-2'
    '<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="en"><head><meta content'

Good! Everything is going as planned.

Also on my laptop, I've Factor, and it has an easy-to-use wgetter in the standard library. Here's `http-get` working on a "normal" WiFi network.

    IN: scratchpad "http://google.com" http-get nip
    
    --- Data stack:
    "<!doctype html><html itemscope=\"\" itemtype=\"http://schema.org..."

Success!

Well, no. `http-get` on my phone's hotspot:

    IN: scratchpad "http://google.com" http-get nip
    
    --- Data stack:
    "\x1f\b\0\0\0\0\0\0\x03Å<ëzÛ¶ÿÏSÐH+K+\"u\x17eÚ&iâÓ¤Ik§i7Íú\x03IHbÄIÊ²#ë]öQw\x06\0..."

Uh.<br>And it's not just Google. `http-get`ting Stack Overflow, or any other website over my phone's network gives rather similar results.

Printing that string:

[![enter image description here][1]][1]

...

No? Ah, well, OK.

---

Factor is 100% UTF-8 by default. `ISO-8859` should be translatable to UTF-8, and indeed, it is when not using my phone's internet.

I know mobile service providers have a reputation of injecting Bad Things into served content. But if the encoding's the same, and Python treats them the same, and Python says they have the same encoding... what's going on here?

---

Factor is `HEAD`. Python is 3.5. Laptop is Ubuntu 15.10, Android is 5.1.something, and probably most importantly, my mobile service provider is StraightTalk. 

As the Python demonstration shows, I don't normally experience issues with page content.

  [1]: http://i.stack.imgur.com/GaT79.png

**Comment by mrjbq7** [comment]
What are the headers when retrieved directly and when retrieved using mobile hotspot?

You can get it with ``curl``:

https://davidwalsh.name/curl-headers

Or with Factor:

```factor
"http://google.com" http-get over header>> >alist simple-table.
```

**Comment by catb0t** [comment]
Aha, I didn't think of that. 

The outstanding one is: `content-encoding`. On normal WiFi, this is `Accept-Encoding`, and on the hotspot it's `gzip`, which explains the compression.

`requests` ungzips, and so does my browser, but Factor's raw. This is very interesting. 
Now I guess the question changes to, since clearly the content is being GZIP'd for size, how can I ungzip it? 

I don't see any `gzip`-y support in the standard library. Perhaps if someone writes such a library, it can be incorporated and `http-get` could decode content if the headers say it's GZIP'd?

**Comment by mrjbq7** [comment]
That sounds like it.  We have some code for it, but need to finish and integrate with ``http.client``.

I wonder if (for now) there is a header we can send to indicate we don't support gzip.

**Comment by catb0t** [comment]
[What encoding string tells a web server NOT to send gzip content?](http://stackoverflow.com/a/12321585/4532996)

Seems like there is a header option that unequivicoally disallows it: `'accept-encoding': 'gzip;q=0,deflate,sdch'`.

**Comment by mrjbq7** [comment]
Good to know.  That's a workaround if this is a blocker for you:

```factor
: my-http-get ( url -- response data )
    <get-request>
        "gzip;q=0,deflate,sdch" "accept-encoding" set-header
    http-request ;
```

**Comment by mrjbq7** [comment]
(Also, I'd be curious if that works for you.)

**Comment by jonenst** [comment]
Also add support for brotli ?

https://en.m.wikipedia.org/wiki/Brotli

https://blog.cdn77.com/cdn77-now-supports-brotli/
Le 26 mars 2016 3:09 AM, "John Benediktsson" <notifications@github.com> a
écrit :

**Comment by catb0t** [comment]
@mrjbq7 No, in the case of my network provider being the one doing the gzipping, the header doesn't have any effect. However, it might be a good idea to add some sort of check in `http-get` for what `content-encoding` is, and ungzip if it's gzipped.
